### PR TITLE
experiment: refactor vault to API of virtual objects

### PR DESCRIPTION
**Pull request by turadg**
_no ticket_

## Description

For consideration of doing before Restival until we can fully virtualized the run-protocol objects. This introduces a `nonvirtualStore` that has a `defineKind` that matches the same API but without the virtualization abiltiies.

### CODE REVIEW

File: packages/run-protocol/src/vaultFactory/vault.js [hunk 1]
```diff
@@ -99,29 +72,70 @@ const makeOuterKit = inner => {
...
-  // CONSTANTS
-  const collateralBrand = manager.getCollateralBrand();
-  const { brand: runBrand } = runMint.getIssuerRecord();
-
-  // STATE
   const { zcfSeat: liquidationZcfSeat, userSeat: liquidationSeat } =
     zcf.makeEmptySeatKit(undefined);
+  return harden({
+    // construction params
+    zcf,
```
Comment (dtribble) [inline packages/run-protocol/src/vaultFactory/vault.js:113]: shouldn't be persistent state

File: packages/run-protocol/src/vaultFactory/vault.js [hunk 2]
```diff
@@ -99,29 +72,70 @@ const makeOuterKit = inner => {
...
+    manager,
+    managerNotifier,
+    idInManager,
+    runMint,
```
Comment (dtribble) [inline packages/run-protocol/src/vaultFactory/vault.js:117]: runMint, priceAuthority, etc should probably come from manager

File: packages/run-protocol/src/vaultFactory/vault.js [hunk 3]
```diff
@@ -99,29 +72,70 @@ const makeOuterKit = inner => {
...
+    // collateral back. If that happens, the issuer for the RUN will be dead,
+    // so their loan will be worthless.
+    vaultSeat: zcf.makeEmptySeatKit().zcfSeat,
+  });
+};
+
+/**
+ * @param {VaultState} state
+ */
+const actualize = state => {
```
Comment (dtribble) [inline packages/run-protocol/src/vaultFactory/vault.js:137]: `actualize` is a really terrible name :)

File: packages/run-protocol/src/vaultFactory/vault.js [hunk 4]
```diff
@@ -147,50 +161,44 @@ export const makeInnerVault = (
...
-  /**
-   * Snapshot of the debt and compounded interest when the principal was last changed
-   *
-   * @type {{run: Amount, interest: Ratio}}
-   */
-  let debtSnapshot = {
-    run: AmountMath.makeEmpty(runBrand, 'nat'),
-    interest: manager.getCompoundedInterest(),
-  };
+  let outerUpdater;
```
Comment (dtribble) [inline packages/run-protocol/src/vaultFactory/vault.js:172]: shouldn't this be persistent state?

File: packages/run-protocol/src/vaultFactory/vault.js [hunk 5]
```diff
@@ -147,50 +161,44 @@ export const makeInnerVault = (
...
 
   /**
    * Called whenever the debt is paid or created through a transaction,
    * but not for interest accrual.
    *
-   * @param {Amount} newDebt - principal and all accrued interest
+   * @param {Amount<NatValue>} newDebt - principal and all accrued interest
```
Comment (dtribble) [inline packages/run-protocol/src/vaultFactory/vault.js:178]: <NatValue> is going to get tiresome real quick, given that it's pervasive for fungible tokens.

File: packages/run-protocol/src/vaultFactory/vault.js [hunk 6]
```diff
@@ -99,29 +72,70 @@ const makeOuterKit = inner => {
...
+    manager,
+    managerNotifier,
+    idInManager,
+    runMint,
+    priceAuthority,
+    // calculated
+    interestSnapshot: manager.getCompoundedInterest(),
+    liquidationSeat,
+    liquidationZcfSeat,
+    runSnapshot: AmountMath.makeEmpty(runMint.getIssuerRecord().brand, 'nat'),
```
Comment (dtribble) [inline packages/run-protocol/src/vaultFactory/vault.js:123]: we slid into naming things for "run".  This should be `debtSnapshot` or `debt`

File: packages/run-protocol/src/vaultFactory/vault.js [hunk 7]
```diff
@@ -227,25 +235,31 @@ export const makeInnerVault = (
...
   const assertVaultHoldsNoRun = () => {
+    const { vaultSeat } = state;
     assert(
       AmountMath.isEmpty(getRunAllocated(vaultSeat)),
       X`Vault should be empty of RUN`,
     );
   };
 
   const maxDebtFor = async collateralAmount => {
+    const { manager, priceAuthority } = state;
```
Comment (dtribble) [inline packages/run-protocol/src/vaultFactory/vault.js:262]: cute!

File: packages/run-protocol/src/vaultFactory/vault.js [hunk 8]
```diff
@@ -273,19 +287,24 @@ export const makeInnerVault = (
...
+    const { manager } = state;
     /** @type {VaultUIState} */
     return harden({
       // TODO move manager state to a separate notifer https://github.com/Agoric/agoric-sdk/issues/4540
       interestRate: manager.getInterestRate(),
       liquidationRatio: manager.getLiquidationMargin(),
-      debtSnapshot,
+      debtSnapshot: {
+        interest: state.interestSnapshot,
+        run: state.runSnapshot,
```
Comment (dtribble) [inline packages/run-protocol/src/vaultFactory/vault.js:306]: need to rename `run`

File: packages/run-protocol/src/vaultFactory/vault.js [hunk 9]
```diff
@@ -386,7 +405,7 @@ export const makeInnerVault = (
     seat.exit();
     burnSeat.exit();
     assignPhase(VaultPhase.CLOSED);
-    updateDebtSnapshot(AmountMath.makeEmpty(runBrand));
+    updateDebtSnapshot(AmountMath.makeEmpty(runBrand, 'nat'));
```
Comment (dtribble) [inline packages/run-protocol/src/vaultFactory/vault.js:408]: what is 'nat'?  This is not the pattern we use elssewhere

File: packages/run-protocol/src/vaultFactory/vault.js [hunk 10]
```diff
@@ -689,18 +706,35 @@ export const makeInnerVault = (
...
     assertVaultIsOpen();
     seat.exit();
+
     // eslint-disable-next-line no-use-before-define
-    return setupOuter(innerVault);
+    const { transferInvitationHook, updater } = setupOuter(innerVault);
+    outerUpdater = updater;
+    updateUiState();
+
+    return transferInvitationHook;
```
Comment (dtribble) [inline packages/run-protocol/src/vaultFactory/vault.js:733]: This is named wrong.  This method is "(makeTransferInvitation)Hook", not "make(TransferInvitationHook)"

File: packages/run-protocol/src/vaultFactory/vault.js [hunk 11]
```diff
@@ -729,6 +763,7 @@ export const makeInnerVault = (
   return innerVault;
 };
 
-/**
- * @typedef {ReturnType<typeof makeInnerVault>} InnerVault
- */
+export const makeInnerVault = defineKind('innerVault', init, actualize);
```
Comment (dtribble) [inline packages/run-protocol/src/vaultFactory/vault.js:766]: I think outervault also needs to be durable

**Review by dtribble — COMMENTED** [review]
WIP but great start

File: packages/run-protocol/src/vaultFactory/vault.js [hunk 12]
```diff
@@ -209,11 +217,11 @@ export const makeInnerVault = (
   const getDebtAmount = () => {
     // divide compounded interest by the snapshot
     const interestSinceSnapshot = multiplyRatios(
-      manager.getCompoundedInterest(),
-      invertRatio(debtSnapshot.interest),
+      state.manager.getCompoundedInterest(),
```
Comment (dtribble) [inline packages/run-protocol/src/vaultFactory/vault.js:220]: use the const pattern to extract state?

File: packages/run-protocol/src/vaultFactory/vault.js [hunk 13]
```diff
@@ -99,29 +72,70 @@ const makeOuterKit = inner => {
...
   priceAuthority,
 ) => {
-  // CONSTANTS
-  const collateralBrand = manager.getCollateralBrand();
-  const { brand: runBrand } = runMint.getIssuerRecord();
-
-  // STATE
   const { zcfSeat: liquidationZcfSeat, userSeat: liquidationSeat } =
     zcf.makeEmptySeatKit(undefined);
+  return harden({
```
Comment (dtribble) [inline packages/run-protocol/src/vaultFactory/vault.js:111]: I'm surprised this should be`harden`ed given that it represents mutable state.

**Review by dtribble — CHANGES_REQUESTED** [review]
A few more comments. Between these two reviews, the explicit typing for NatValue and the naming of the hooks need to be addressed. The basic transforamtion all looks great.

**Comment by turadg** [comment]
Experiment concluded. Most of this made it into another PR

**Pull request closed without merging**